### PR TITLE
Added test examples for SBOMs in multiple formats

### PR DESCRIPTION
**Pull request by wei-deepbits**
Hi all,

I've opened this pull request to contribute test examples for SBOM conversion. These examples are used to verify the correctness of our SBOM translate tool by testing the conversion process.

The test examples include SBOMs for the sample software in three different formats: CycloneDX-1.4 JSON, SPDX-2.3 JSON, and SPDX-2.3 tag-value.

I appreciate your review and feedback on these additions.

Thank you,

Wei Song
Deepbits.com

### CODE REVIEW

**Comment by jspeed-meyers** [comment]
Cool @wei-deepbits! One question I have before merging is: is there any documentation about how these SBOMs were created? I want to make sure that future users of this codebase fully understand these SBOMs, especially if they become part of the test suite. In short, adding some documentation about how these SBOMs were created would be useful.

**Comment by wei-deepbits** [comment]
Hi @jspeed-meyers, I have uploaded a file named README.md to provide an explanation of how these SBOMs were created. You can find it in the examples/tests directory.

**Comment by jspeed-meyers** [comment]
@wei-deepbits, ahhh. Cool Thank you for adding this README.

I noticed this part:

> Step 1: Generating a Configuration File
To begin, we use ChatGPT to generate a configuration file that contains randomly selected packages for various programming languages. In the case of a Python package, we generate a "poetry.lock" file located in the "PYTHON/" folder.

Are you confident that ChatGPT has created valid configuration files for each programming language? Did you manually check? If you are, then I support merging this.

If you're not sure if these files are valid, another technical possibility is to use a real-world example of each type of package manager metadata file (what you called a "configuration") file. You could select these package manager metadata files only from real-world, popular open source projects. Just a thought!

One other broad thought on something you mentioned:

> These examples are used to verify the correctness of our SBOM translate tool by testing the conversion process.

IIUC, I'm personally not sure that these files can be directly used to test the correctness of the translation process. Because there is inevitably loss between different formats, it might not be possible to use these documents to test, for instance, if the Python app SPDX SBOM document generated via `trivy` will result, once translated by this tool, into the Python app CycloneDX SBOM document generated by `trivy`. In other words, these SBOMs are helpful, I think, but likely can't be used directly for our testing purpose. Am I making sense? Perhaps I am mistaken in some assumption or logic too. Regardless, I think these SBOMs will be helpful for testing! Thank you for this work!

**Comment by puerco** [comment]
Mmh I don't know about these. We need to start scaffolding some conformance tests, but it looks like some of these SBOMs may not be very useful. We could start treating them as edge cases to handle the limits of the tool, but for regular development and testing I think they are not very good. Some examples from he SPDX ones:

First, the SPDX documents are not valid SPDX:

```
This SPDX Document is not valid due to:
	License list version does not match the pattern M.N
	Missing required package files for GO/go.sum
	Missing required package verification code for package GO/go.sum
```

The go SPDX SBOM is missing all data, it just lists some made-up package without anything useful:

```json
{
      "name": "GO/go.sum",
      "SPDXID": "SPDXRef-Filesystem-145b8edeffbf5e91",
      "downloadLocation": "NONE",
      "copyrightText": "",
      "attributionTexts": [
        "SchemaVersion: 2"
      ],
      "primaryPackagePurpose": "SOURCE"
    }

```

The  java SPDX example contains the same package but judging from the name it should be a file. Buit again note that there are no hashes, no identifiers, and nothing useful:

```json

    {
      "name": "JAVA/pom.xml",
      "SPDXID": "SPDXRef-Filesystem-60c263cac2575773",
      "downloadLocation": "NONE",
      "copyrightText": "",
      "attributionTexts": [
        "SchemaVersion: 2"
      ],
      "primaryPackagePurpose": "SOURCE"
    },
```

**Comment by jspeed-meyers** [comment]
@wei-deepbits:

Like @puerco mentions, making sure these SBOMs are valid by their respective specifications (SPDX or CycloneDX) is indeed important.

I think merging this PR to keep these SBOMs around for edge case testing could be worthwhile, but I think using better vetted, more standard, non-ChatGPT-derived SBOMs for the main test cases might be the right move.

**Comment by wei-deepbits** [comment]
Got it. I will use real-world package manager metadata files and validate the generated SBOMs.

**Comment by wei-deepbits** [comment]
Hi @jspeed-meyers and @puerco. I have removed metadata files generated by ChatGPT and added a new script to collect real metadata files from the awesome-XXX repository.

File: examples/tests/README.md
Comment (jspeed-meyers) [inline examples/tests/README.md:38]: Is it actually `download_metadata.py`?

File: examples/tests/README.md
Comment (jspeed-meyers) [inline examples/tests/README.md:39]: I noticed there are ZIP files in this PR but but not folders? Can you please help me understand? Are they the same thing really?

**Comment by jspeed-meyers** [comment]
Cool, @wei-deepbits! Question: Have you already collected these SBOMs and that's what is in the ZIP files? Or are those ZIP files something different?

My own 2 cents: I think these SBOMs could indeed be useful for the purposes of the test suite. Looking forward to merging this soon.

File: examples/tests/README.md
Comment (wei-deepbits) [reply examples/tests/README.md:38]: Yes. I will correct this typo.

File: examples/tests/README.md
Comment (wei-deepbits) [reply examples/tests/README.md:39]: In my latest push, I deleted two folders containing numerous files and added two zip files. I will update the README.md

**Comment by wei-deepbits** [comment]
Hi @jspeed-meyers, these two zip files contain partial results. There are thousands of repositories that need to be downloaded in the "download_metadata.py" file. It is not yet finished. You can run the script to download more metadata files.

**Comment by wei-deepbits** [comment]
Hi @jspeed-meyers, I have addressed the issues you mentioned and made the necessary corrections.

**Comment by jspeed-meyers** [comment]
@puerco and @houdini91, any thoughts? It seems to me that adding these SBOMs for potential test case usage later would be worthwhile. Any questions or thoughts before merging?

Nice work, @wei-deepbits.

**Comment by puerco** [comment]
@wei-deepbits can you explain a little bit more how we'll be using the scraper and the data? Do we need to have both checked into the repo?  Is this intended to be used in tests? I understand from the README what it does but I'm trying to understand how we will be using the scraper.

**Comment by wei-deepbits** [comment]
@puerco, Thank you for your review. You're right, we don't need to include both the script and data in the repository. I ran the script beforehand and generated data files for all SBOMs. As the data files were substantial in size, I removed them from the repository and made them available for download on a shared Google Drive. The repository now only contains the essential script files. The generated SBOMs are intended for use in tests, as they are generated from the same real-world repository in three formats. I believe these SBOMs will be useful for testing the accuracy of our translator.

**Comment by wei-deepbits** [comment]
Hi, @puerco. Would you kindly consider reviewing the new commit and sharing your thoughts? Thanks.

**Comment by jspeed-meyers** [comment]
@wei-deepbits: @puerco is traveling on vacation this week, though he might pop his head onto GitHub in a free moment or two. I wanted you to know that he isn't ignoring the commit--he's just away from his computer!

File: examples/tests/crawl_and_generate_SBOM.py
Comment (jspeed-meyers) [inline examples/tests/crawl_and_generate_SBOM.py:1]: This is a nit. [PEP 8](https://peps.python.org/pep-0008/#imports) in Python states this about imports:

> Imports should be grouped in the following order:
Standard library imports.
Related third party imports.
Local application/library specific imports.
You should put a blank line between each group of imports.

I *think* the correct ordering of imports would be this:

```suggestion
import json
import os
import re
import subprocess
import tempfile
from urllib.parse import urlparse

import requests
from git import Repo
```

File: examples/tests/crawl_and_generate_SBOM.py
Comment (jspeed-meyers) [inline examples/tests/crawl_and_generate_SBOM.py:44]: It's not obvious to me what this function "validates" a CycloneDX SBOM. Consider adding a comment.

File: examples/tests/requirements.txt
Comment (jspeed-meyers) [inline examples/tests/requirements.txt:1]: Do you want to pin versions?

**Review by jspeed-meyers — COMMENTED** [review]
This looks good to me save a few nits and questions.

**Comment by wei-deepbits** [comment]
Hello, @jspeed-meyers! I appreciate your reviews and suggestions.

I've made some updates based on your feedback. Firstly, I rearranged the import orders. Secondly, I renamed the function from "validate_cyclonedx_sbom" to "is_library_in_sbom" to better reflect its purpose. The main goal of this function is to address the issue where more than half of the SBOMs generated by Syft are empty and contain no dependency packages. To resolve this, I added the function to filter out SBOMs that have no packages, ensuring that only those with at least one package are retained. Furthermore, I've taken pinned versions in the requirements.txt file. 

Thanks again for your valuable input. Let me know if you have any further suggestions or questions!

**Review by jspeed-meyers — APPROVED** [review]
@wei-deepbits, great work! That all makes sense to me. Thank you.

Let's get one more review beyond myself and then merge this!

File: examples/tests/crawl_and_generate_SBOM.py
Comment (manifestori) [inline examples/tests/crawl_and_generate_SBOM.py:11]: It would be beneficial to allow the user to specify the `SBOM_OUTPUT` directory. This could be done by checking for an environment variable and fallback to "./SBOM".

File: examples/tests/crawl_and_generate_SBOM.py
Comment (manifestori) [inline examples/tests/crawl_and_generate_SBOM.py:89]: Would it be a good idea to output the SBOM to `stdout` if no SBOM_OUTPUT or dedicated flag for a file path is defined? (similar to `syft` and other generator behavior)

**Review by manifestori — APPROVED** [review]
I'm not a python expert, but LGTM :)

**Comment by wei-deepbits** [comment]
Hi @manifestori , thanks for your valuable input. I've added the parse_arguments() function to handle command-line arguments for the SBOM generation script. Now users can easily specify the output folder for SBOMs using either the -o or --output options. If the SBOM_OUTPUT environment variable is set, the script will utilize that value. Otherwise, it defaults to "./SBOM/". Regarding your second suggestion, I think that outputting the SBOMs to stdout might be overwhelming, especially with thousands of SBOMs. For now, let's continue with writing the SBOMs to files. Let me know if you have any more ideas!

**Comment by jspeed-meyers** [comment]
@wei-deepbits, I'll let @manifestori take one more look. If Ori is happy with the changes, then let's merge!

Thank you, @wei-deepbits, for the diligence on this. Having a useful set of SBOMs for testing is really important. Thank you.

And thank you, @manifestori, for the review.

**Comment by manifestori** [comment]
Looks great, well done! 💯

**Comment by wei-deepbits** [comment]
@jspeed-meyers @manifestori and @puerco, Your feedback is truly valuable and deeply appreciated. Thank you!

**Comment by jspeed-meyers** [comment]
Great work, @wei-deepbits!

**Pull request merged**
